Add cluster definition for smoke-co-alarm

ericzijian1994 opened the pull request:
This PR adds an implementation of a Smoke CO Alarm.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=26474) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=26474) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=26474) it.</sub>

bzbarsky-apple commented on examples/smoke-co-alarm-app/silabs/efr32/.gn:
Please split out the XML additions and core regeneration into a separate PR.  They need to be _very_ carefully reviewed, and are too easy to lose in all the other changes here.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml:
```diff
@@ -0,0 +1,38 @@
+    <domain>Measurement &amp; Sensing</domain>
+    <description>Attributes and commands for configuring the measurement of concentration, and
+      reporting concentration measurements.</description>
+    <code>0x040C</code>
```
This is the Carbon Monoxide Concentration Measurement cluster (not clearly labeled so in the spec, sadly), not the generic Concentration Measurement.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml:
```diff
@@ -0,0 +1,38 @@
+    <define>CONCENTRATION_MEASUREMENT_CLUSTER</define>
+    <client tick="false" init="false">true</client>
+    <server tick="false" init="false">true</server>
+    <globalAttribute side="either" code="0xFFFD" value="1" />
```
Spec says 3, no?  Going to stop here; please break out the XML changes into a separate PR so we can get the XML aligned to the spec without noise from all the other bits.

ericzijian1994 replied:
Clear, will open a new PR for XML changes.

bzbarsky-apple commented:
@ericzijian1994 First, I can handle the availability.yaml bits, so no need to touch them.

Second, why are you closing this?  That's going to make the review comment tracking very hard.

bzbarsky-apple commented:
Ah, I guess you'll do separate PRs for the XML and the other bits, right.  In that case, closing probably makes sense if you mention this one in the others.

ericzijian1994 commented:
@bzbarsky-apple, I have updated the content.

The major new folders are:

1. examples/smoke-co-alarm-app
2. src/app/clusters/smoke-co-alarm-server

In addition, I updated the smoke-co-alarm.xml. #26705

bzbarsky-apple commented:
> In addition, I updated the smoke-co-alarm.xml. https://github.com/project-chip/connectedhomeip/issues/26705

At risk of sounding like a broken record: please put the XML changes in a separate pr @ericzijian1994

ericzijian1994 commented:
@bzbarsky-apple, I have recommitted the code.

ericzijian1994 commented:
@bzbarsky-apple, now using the new XML that was merged into the master.

bzbarsky-apple commented on examples/smoke-co-alarm-app/silabs/efr32/src/ZclCallbacks.cpp:
```diff
@@ -0,0 +1,64 @@
+
+    if (clusterId == SmokeCoAlarm::Id && attributeId == SmokeCoAlarm::Attributes::TestInProgress::Id)
+    {
+        AlarmMgr().StartSelfTesting();
```
What if TestInProgress is being set to false?

bzbarsky-apple commented on examples/smoke-co-alarm-app/silabs/efr32/src/ZclCallbacks.cpp:
```diff
@@ -0,0 +1,64 @@
+    }
+    else if (clusterId == Identify::Id)
+    {
+        ChipLogProgress(Zcl, "Identify attribute ID: " ChipLogFormatMEI " Type: %u Value: %u, length %u",
+                        ChipLogValueMEI(attributeId), type, *value, size);
```
This log line only makes sense if size is known to be 1.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -0,0 +1,354 @@
+        }
+        else
+        {
+            success = SetTestInProgress(commandPath.mEndpointId, true);
```
Spec says ExpressedState should be set to Testing when test starts and then cleared when test stops.  Is the application responsible for doing that?  If so, that should be documented in the header API documentation.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -0,0 +1,354 @@
+
+    if (!success)
+    {
+        ChipLogError(Zcl, "Failed to read SmokeCOAlarm attribute: attribute=0x%" PRIx32 ", status=0x%x", attributeId,
```
```suggestion
        ChipLogError(Zcl, "Failed to read SmokeCOAlarm attribute: attribute=" ChipLogFormatMEI ", status=0x%x", ChipLogValueMEI(attributeId),
```

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -0,0 +1,354 @@
+
+    if (!success)
+    {
+        ChipLogError(Zcl, "Failed to write SmokeCOAlarm attribute: attribute=0x%" PRIx32 ", status=0x%x", attributeId,
```
```suggestion
        ChipLogError(Zcl, "Failed to write SmokeCOAlarm attribute: attribute=" ChipLogFormatMEI ", status=0x%x", ChipLogValueMEI(attributeId),
```

ericzijian1994 commented:
@bzbarsky-apple I've modified them, added documentation to the headers, and submitted a spec issue.

ericzijian1994 commented:
@bzbarsky-apple I have revised the conflicting content, please review, thanks.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:
```diff
@@ -0,0 +1,178 @@
+#include <protocols/interaction_model/StatusCode.h>
+
+#ifndef SMOKE_CO_ALARM_SERVER_ENDPOINT
+#define SMOKE_CO_ALARM_SERVER_ENDPOINT 1
```
What is this define and why is it here?  It seems completely unused.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @param endpointId ID of the endpoint
+     * @param newExpressedState new expressed state
+     * @return true on success, false on failure
```
What I meant is that one comment like "For all the functions below, this is what the return value means" would be OK.  This comment just documents a single function...

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -0,0 +1,460 @@
+                                                       const Commands::SelfTestRequest::DecodableType & commandData)
+{
+    return SmokeCoAlarmServer::Instance().HandleRemoteSelfTestRequest(commandObj, commandPath,
+                                                                      emberAfPluginSmokeCoAlarmSelfTestRequestCommand);
```
If we are hardcoding this anyway, why not have HandleRemoteSelfTestRequest call this function directly instead of passing the pointer, etc?  Either way, though.

bzbarsky-apple commented on examples/smoke-co-alarm-app/silabs/src/ZclCallbacks.cpp:
```diff
@@ -0,0 +1,66 @@
+
+    if (clusterId == SmokeCoAlarm::Id && attributeId == SmokeCoAlarm::Attributes::ExpressedState::Id)
+    {
+        SmokeCoAlarm::ExpressedStateEnum expressedState = *(reinterpret_cast<SmokeCoAlarm::ExpressedStateEnum *>(value));
```
Please `static_assert(sizeof(SmokeCoAlarm::ExpressedStateEnum) == 1, "Wrong size");` here, since you are assuming that.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -0,0 +1,460 @@
+        }
+        else
+        {
+            success = opHandler(endpointId);
```
Again: what is responsible for setting ExpressedState to "Testing"?  If it's the application then:

1. The example app being added here should actually do it!
2. That needs to be very clearly documented on the callback declaration.

ericzijian1994 replied:
I have now changed to set ExpressedState at the server, so I have added the ExpressedState priority implementation.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -0,0 +1,354 @@
+    bool success =
+        SetAttribute(endpointId, Attributes::EndOfServiceAlert::Id, Attributes::EndOfServiceAlert::Set, newEndOfServiceAlert);
+
+    if (success && (newEndOfServiceAlert == EndOfServiceEnum::kExpired))
```
Spec issue is https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/7091

Next time, @ericzijian1994 please mention the link when filing a spec issue?

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:
```diff
@@ -158,6 +168,11 @@ class SmokeCoAlarmServer
+    /**
+     * @brief Expression status record values
+     */
+    int expressedStateMask = 1;
```
Why is this signed?  and why is it not named with "m...."?

And the description of this member does not make sense...

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:
```diff
@@ -102,6 +104,15 @@ class SmokeCoAlarmServer
     inline bool SupportsCOAlarm(chip::EndpointId endpointId) { return GetFeatures(endpointId).Has(Feature::kCoAlarm); }
 
+    /**
+     * @brief ExpressedState priority, the priority order of conditions is determined by the manufacturer
```
If it's determined by the manufacturer, why is it hardcoded here?  This does not make sense.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -55,8 +55,21 @@ SmokeCoAlarmServer & SmokeCoAlarmServer::Instance()
+    expressedStateMask |= (1 << to_underlying(expressedState));
+
+    // Get the highest priority alarm
+    for (ExpressedStateEnum i : expressedStatePriority)
```
This logic needs to be in the app, not in the cluster.  Why are these changes being made?

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -387,14 +443,23 @@ bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj
         {
-            success = opHandler(endpointId);
-            if (!success)
+            success = SetExpressedState(endpointId, ExpressedStateEnum::kTesting);
```
Is this why the priority stuff was added? 

My understanding of the spec is that expressed state unconditionally changes to kTesting when this command is successful, which is not what the priority stuff implements, right?

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -387,14 +443,23 @@ bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj
+            {
+                success = emberAfPluginSmokeCoAlarmSelfTestRequestCommand(endpointId);
+
+                UnsetExpressedState(endpointId, ExpressedStateEnum::kTesting);
```
This is wrong.  The expressed state needs to change back _when the test is done_.  Which means the app needs to do it.

bzbarsky-apple commented on src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:
```diff
@@ -387,14 +443,23 @@ bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj
+                success = emberAfPluginSmokeCoAlarmSelfTestRequestCommand(endpointId);
+
+                UnsetExpressedState(endpointId, ExpressedStateEnum::kTesting);
+                SetTestInProgress(endpointId, false);
```
Again, the app needs to do that.  We don't know when the test is done.

If we want to make this easier, we could have a function for "test is done, here is the new expressed state" that the app would call, but all that would do is SetExpressedState and SetTestInProgress(false).

ericzijian1994 replied:
OK, change to APP to set the state.

bzbarsky-apple commented on examples/smoke-co-alarm-app/silabs/src/SmokeCoAlarmManager.cpp:
```diff
@@ -37,10 +37,29 @@ CHIP_ERROR SmokeCoAlarmManager::Init()
-    SILABS_LOG("End self-testing!");
+    if (success)
+    {
+        success = SmokeCoAlarmServer::Instance().SetExpressedState(1, ExpressedStateEnum::kTesting);
```
Again: _this_ part should live in the cluster, because all applications will need to do it.

The app knows what its expressed state is (the attr store is not the source of truth for that), and once the self-test is done should update the expressed state to the right value.  But setting it to kTesting should happen in the cluster itself, I would think.

ericzijian1994 replied:
I have moved this part to cluster.
